File: Little-book-of-semaphores/3.3/Rendezvous.java
import java.util.concurrent.Semaphore;

public class Rendezvous {
    static Semaphore mutex1 = new Semaphore(0);
    static Semaphore mutex2 = new Semaphore(0);

    public static void main(String[] args) {

        Thread threadA = new ThreadA();
        Thread threadB = new ThreadB();

        threadA.start();
        threadB.start();
    }

    static class ThreadA extends Thread {
        public void run() {
            try {
                System.out.println("action a1");
                mutex1.release();
                mutex2.acquire();
                System.out.println("action a2");
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    static class ThreadB extends Thread {
        public void run() {
            try {
                System.out.println("action b1");
                mutex2.release();
                mutex1.acquire();
                System.out.println("action b2");
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }
}

File: Little-book-of-semaphores/3.3/Rendezvous.py
from threading import Thread
from threading import Lock

signalA1 = Lock()
signalA2 = Lock()

signalA1.acquire()
signalA2.acquire()

def a():
    print('action a1')
    signalA1.release()
    signalA2.acquire()
    print('action a2')

def b():
    print('action b1')
    signalA2.release()
    signalA1.acquire()
    print('action b2')

threadA = Thread(target=a)
threadB = Thread(target=b)

threadA.start()
threadB.start()